### PR TITLE
fix tensor layout validation for conv2d 

### LINKED ISSUE #47047: [ops]: conv2d may internally construct tensor layout object with invalid alignment
### Component / Area

ops

### Issue Type (optional)

Other

### Observed

`conv2d` operation may call the `TensorLayout::with_memory_config` with invalid memory configuration layout. When validation is added to the `with_memory_config`, the sanity test execution might fail with exception indicating the conv2d constructs invalid tensor layout. 

### Expected

Memory config passed to the conv2d is valid for all cases and does not rely on the incorrect implementation details of `with_memory_config` to work. 

### 1. Steps (exact commands)

Adding alignment validation call to the `with_memory_config` reveals the invalid tensory layout construction. 
 
- on branch [martemov/main-layout-alignment-validation](https://github.com/tenstorrent/tt-metal/tree/refs/heads/martemov/main-layout-alignment-validation) with commit https://github.com/tenstorrent/tt-metal/commit/41679ed4b7ffc4a711300bc6bcf8f71275f1c458
- test `tests/ttnn/unit_tests/operations/conv/test_conv2d.py::test_conv_features[output_layout=Layout.TILE-math_fidelity=MathFidelity.HiFi4-filter=3-padding=(1, 2, 2, 3)-packer_l1_acc=True-fp32_accum=True-input_dtype=DataType.BFLOAT16-output_dtype=DataType.BFLOAT8_B-output_channels=128-input_channels=128-input_height=32-input_width=32-shard_layout=TensorMemoryLayout.BLOCK_SHARDED-config=None-batch_size=2-stride=2-device_params={'l1_small_size': 16384}]`

### 2. Input data / link or description

The issue is reproducible by the test `tests/ttnn/unit_tests/operations/conv/test_conv2d.py::test_conv_features` on the provided branch. 

### 3. Frequency

always

### 1. Software Versions

Ubuntu, firmware 5.2.0

### 2. Hardware Details

P150b, may be reproduced on other cards. 

### Is this a regression?

No

### Regression Details

_No response_

### Logs & Diagnostics

```
/project/tt_metal/impl/tensor/spec/layout/page_config.cpp:178: physical_shard_width % width_alignment == 0
info:
Alignment mismatch for sharded tensor: Expected physical shard shape 16 to be aligned to 128 along the width for Row Major layout.
backtrace:
 --- /opt/venv/lib/python3.10/site-packages/ttnn/build/lib/_ttnncpp.so(+0x6f8ce8) [0x7f3f50a95ce8]
 --- tt::tt_metal::RowMajorPageConfig::validate_alignment(tt::tt_metal::Alignment const&, tt::tt_metal::DataType, tt::tt_metal::MemoryConfig const&) const
 --- tt::tt_metal::TensorLayout::with_memory_config(tt::tt_metal::MemoryConfig) const
```

### Priority

None

### Impact

This issue is blocking full implementation of https://github.com/tenstorrent/tt-metal/issues/42827 and removal of the `TensorLayout::with_memory_config`.

### PR DESCRIPTION
**Pull request by ctr-martemovTT**
### PR Category

fix: https://github.com/tenstorrent/tt-metal/issues/47047

### Summary

conv2d may call the ttnn::pad before moving the tensor to device, which causes issues with the tensor alignment and shard size. Specifically, the failing test mentioned in the linked issue failed because

- input tensor was padded to last dimension 128
- the core grid had 8 cores on this dimension, resulting in the shard last dimension 16
- validation for the tensor layout expects `physical_shard_width % width_alignment == 0` for row-major block-sharded tensors

The PR changes the on-device branch for the conv2d utils sharding logic to comply with the spec validation rules. 

### Notes for reviewers

This is necessary for the follow-up PR for https://github.com/tenstorrent/tt-metal/issues/42827 -- to remove the TensorLayout::with_memory_config and replace it with the direct tensor layout constructor call.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:martemov/main-layout-alignment-validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:martemov/main-layout-alignment-validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:martemov/main-layout-alignment-validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:martemov/main-layout-alignment-validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:martemov/main-layout-alignment-validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:martemov/main-layout-alignment-validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:martemov/main-layout-alignment-validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:martemov/main-layout-alignment-validation)

### CODE REVIEW

**Comment by pavlejosipovic** [comment]
1. This needs a regression tests
2. post passing https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml and bh post-commit
3. why is the perf impact of this change?

**Comment by ctr-martemovTT** [comment]
@pavlejosipovic 

Added tests to verify new branches in the conv2d

### perf tests

- https://github.com/tenstorrent/tt-metal/actions/runs/28504776059/job/84493475501
- https://github.com/tenstorrent/tt-metal/actions/runs/28438176572/job/84274244695

unfortunatetely the main branch for the perf tests is still not fully OK, but the two runs show no performance regressions compared to the state of `main` at the time of triggering the pipeline (e.g. https://github.com/tenstorrent/tt-metal/actions/runs/28562781561). 

### sanity tests 

- https://github.com/tenstorrent/tt-metal/actions/runs/28504228435 green OK

### L2 nightly tests for conv2d

- https://github.com/tenstorrent/tt-metal/actions/runs/28521772386 green OK

### blackhole sanity 

- https://github.com/tenstorrent/tt-metal/actions/runs/28521761850 -- no regressions compared to the `main` branch (e.g. https://github.com/tenstorrent/tt-metal/actions/runs/28567619256)

**Comment by mstaletovicTT** [comment]
PLease post  https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml aswell

**Comment by ctr-martemovTT** [comment]
@mstaletovicTT 

perf-models run 

https://github.com/tenstorrent/tt-metal/actions/runs/28579199786 no regressions compared to main run (e.g. https://github.com/tenstorrent/tt-metal/actions/runs/28574960548)

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp [hunk 1]
```diff
@@ -898,10 +888,79 @@ std::tuple<ttnn::Tensor, ParallelConfig, ParallelConfig> shard_or_reshard_tensor
...
+            };
+
+            // if the input is row-major and the shard configuration is width/block sharded AND the physical shape of
+            // the tensor is not aligned with the shard shape, it is necessary to create an on-device tensor with the
+            // target alignment on-device. This edge case is required because
+            //
+            // - Using `ttnn::pad` changes the alignment on the tensor to its physical size, and the tensor spec
+            //   validation requires that `physical_shard_width % width_alignment == 0`
+            // - `physical_size.width() % page_shape.width() == 0`
+            if (input_tensor.logical_shape()[-1] % input_tensor_sharded_memory_config.shard_spec()->shape[-1] != 0 &&
```
Comment (mstaletovicTT) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp]: I think shard_spec->shape is an std::array, you should not be able to index it like this

Comment (ctr-martemovTT) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp]: **resolved**: changed access to the last dimension to `[1]` -- shape is an `std::array` with two elements.

**Review by riverwuTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp [hunk 2]
```diff
@@ -898,10 +888,79 @@ std::tuple<ttnn::Tensor, ParallelConfig, ParallelConfig> shard_or_reshard_tensor
                 input_tensor = resharded_input_tensor;
             }
         } else {
-            input_tensor = ttnn::to_device(
-                input_tensor, device, (auto_shard_mm ? ttnn::DRAM_MEMORY_CONFIG : input_tensor_sharded_memory_config));
```
Comment (s-jovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:902]: If I am understanding correctly:

1. Alignment per dimension is set by ttnn:pad, creating a ttnn tensor and not padding it leaves alignment as [1] simply. Is this intended? [Example test](https://github.com/tenstorrent/tt-metal/commit/3ada18b0c699b654690291d75e190a059006bbc8) I ran gave following results:
```
BEFORE pad  (width 125, unpadded):
  logical   : [1, 1, 2, 125]
  padded    : [1, 1, 2, 125]
  alignment : [1]

AFTER  pad  (width 125 -> 128):
  logical   : [1, 1, 2, 125]
  padded    : [1, 1, 2, 128]
  alignment : [2, 128]
```

2. If 1 is intended, shouldn't `to_device` be able to handle alignment when converting interleaved tensor to sharded? If so, should the fix be inside i2s op and not conv?

cc @mstaletovicTT

Comment (ctr-martemovTT) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:902]: @s-jovic 

### 1. 

alignment per dimension is the standard way of padding for RM interleaved tensors, https://github.com/tenstorrent/tt-metal/blob/5eeaa3714a3ad894af703167329f7b854cb2cbbf/tt_metal/impl/tensor/spec/layout/tensor_layout.cpp#L37 

### 2.

The main reason for this PR and the focus on the conv2d operation specifically is because the conv2d operation is providing invalid arguments to the `to_device` sharding -- there is a explicit validation rule for `physical_shard_width % width_alignment == 0,` in 

https://github.com/tenstorrent/tt-metal/blob/5eeaa3714a3ad894af703167329f7b854cb2cbbf/tt_metal/impl/tensor/spec/layout/page_config.cpp#L182 

which itself is linked to how RM sharding paging is implemented.

https://github.com/tenstorrent/tt-metal/blob/5eeaa3714a3ad894af703167329f7b854cb2cbbf/tt_metal/impl/tensor/spec/layout/page_config.cpp#L205

So changes in to_device would require dropping the validation, or reworking how sharded RM tensors are handled in some scenarios, to support the conv2d, which seems to be the only operation breaking after the validation is enabled.

Comment (s-jovic) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:902]: But the broken `with_memory_config` function is not directly called by conv2d op. 

Isn't this a valid chain of ops, regardless of conv2d:

```
padded_input = ttnn.pad(input, padding)
sharded_input = ttnn.interleaved_to_sharded(padded_input, memory_config)
```

Comment (ctr-martemovTT) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:902]: @s-jovic 

The call stack in question that was leading to validation failure is 

```
  frame #0: void tt::assert::detail::tt_throw<unsigned int, unsigned int>
  frame #1: tt::tt_metal::RowMajorPageConfig::validate_alignment
  frame #2: tt::tt_metal::TensorLayout::with_memory_config
  frame #3: tt::tt_metal::TensorLayout::with_memory_config
  frame #4: tt::tt_metal::enqueue_write_tensor
  frame #5: tt::tt_metal::to_device
  frame #6: tt::tt_metal::Tensor::to_device
  frame #7: ttnn::operations::core::to_device
  frame #8: ttnn::operations::conv::shard_or_reshard_tensor_if_required
  frame #9: ttnn::operations::conv::conv2d::conv2d_L1
  frame #10: ttnn::conv2d

```

`tt::tt_metal::Tensor::to_device` and the rest of the functions until the `enqueue_write_tensor` directly pass the `memory_config` and tensor value produced in the `shard_or_reshard_tensor_if_required`, so while the conv2d does not *directly* call the validation, the value *produced by the conv2d* is passed unmodified. 


> But the broken `with_memory_config` function is not directly called by conv2d op.
> 
> Isn't this a valid chain of ops, regardless of conv2d:
> 
> ```
> padded_input = ttnn.pad(input, padding)
> sharded_input = ttnn.interleaved_to_sharded(padded_input, memory_config)
> ```

No, at least not with any arbitrary input tensor and any arbitrary memory config. That's the main change caused by enabling pre-existing validation in `with_memory_config`. 

After validation is enabled the move to device is valid as long as the arguments pass the checks. I.e. the important part here is the `to_device` call used in `shard_or_reshard_tensor_if_required`

```c++
ttnn::to_device(input_tensor, device, memory_config);
```

This call is not valid for *any* pair of input tensor + memory config, For RM tensor converted to sharding, the requirement is - `physical_shard_shape[1] % alignment[-1] == 0` 

- `alignment[-1]` is affected by how the pad is called around `to_device` move -- that's why reordering of the pad relevant here, specifically in conv2d. 
- `physical_shard_shape[1]` is directly influenced by the conv2d logic. 

So, with validation enabled, not every single chain of `pad + to_device` is valid, in a sense that not every combination of arguments to `to_device` is valid.

Specifically, this can be reduced even further, as the main focus for the PR was to ensure that `to_device` arguments coform to the pre-existing validation rules (the validation logic itself was not introduced in the PR, only enabled in the `with_memory_config`). `conv2d` is the operation calling `to_device` with the invalid arguments.

Comment (s-jovic) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:902]: My example was wrong sorry, i2s is not called in this path, to_device ends up calling [enqueue write tensor](https://github.com/tenstorrent/tt-metal/blob/main/tt_metal/impl/tensor/tensor_apis.cpp#L98) as you wrote - but in any case shouldn't this function be able to handle padded alignment - why is it forbidden to push padded tensor to device with sharded config? Why would this be invalid combination as you mentioned?

Comment (ctr-martemovTT) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:902]: @s-jovic 

- it's not forbidden in general, but in this specific situation (row-major sharded into the width/block) because this sharding configuration requires storing one page in one row of a shard (page_config.cpp:205) and changing alignment messes it up because every page must start at the `width_alignment` boundary. This must also be true in case of the sharded tensor.

- So in the specific case: the input tensor gets `alignment = 128` (its padded row width), and the block-sharded config has shard width `16`. This can't be stored because `16 % 128 != 0` — with a page of only 16 elements the pages land at offsets 16, 32 which aren't multiples of 128. The RM row of size 128 simply can't be paged into 16-wide shard blocks.

So to reiterate, it is not forbidden to push the *padded* tensor to device with *sharded* config -- it is forbidden to push *padded row-major tensor into the device with block/width sharded config if the alignment of the padded tensor does not match up with the shard width* (i.e. the restriction is not about padded/non-padded or sharded/non-sharded, but rather a much more narrow combination of alignment, sharding, layout etc.)

- the first branch handles the case for the input tensor logical shape mismatch for RM tensor, by creating tensor with the matching alignment 
- the else branch handles the other cases, only triggering the post-device padding in the RM+block/width sharding cases. 

hope this clears up the reason for the validation rules. The main part relevant here relates to the row-major page config in page_config.cpp file.

Comment (s-jovic) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:902]: > if the alignment of the padded tensor does not match up with the shard width

Should there be an API that allows us to prepare this kind of tensor for sharding? Sorry for going back and forth, I am not understanding what is fundamentally wrong with pushing such padded tensor on device with the needed shard spec. If it is easier we can sync on Slack as well, just make sure to include @mstaletovicTT as well since I will be off next week.

Comment (ctr-martemovTT) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:902]: @s-jovic I will reach out to mstaletovicTT to continue the review process, no problem. 

Regarding the sharding tensor and interactions with the page config, I believe the deepwiki summary provides an excellent level of details for these specfic questions https://deepwiki.com/search/why-rowmajorpageconfigvalidate_59ed6e70-ffa9-4385-b930-625a6865d142?mode=deep

specifically "Why RowMajorPageConfig::validate_alignment has the requirement for `physical_shard_width % width_alignment == 0`" and particular edge case regarding small shard size and large tensor alignment (like 16 vs 128 in the case of the failing test).

Comment (mstaletovicTT) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:902]: The question I want to understand is whether it is possible for to_device to always construct a tensor with legal alignement, regardless of the caller.

Because I do not really understand why currently, when doing to_device, we try to inherit the HOST tensor alignement for the DEVICE tensor. Could we not, prior to the transfer, actually calculate the alignement that the tensor would need on device ?

Comment (ctr-martemovTT) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:902]: @mstaletovicTT because it is necessary to know the size and alignment of the host buffer to correctly allocate the device tensor. Specifically, the `allocate_device_buffer` transitively calls into `TensorLayout::compute_physical_shape` that relies on the alignment information to determine the size of the buffer. 

So while it might be theoretically possible to change how host-device tensor moves are done, it is likely require a high-risk high-impact change in this area.

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp [hunk 3]
```diff
@@ -898,10 +888,79 @@ std::tuple<ttnn::Tensor, ParallelConfig, ParallelConfig> shard_or_reshard_tensor
...
+            // the tensor is not aligned with the shard shape, it is necessary to create an on-device tensor with the
+            // target alignment on-device. This edge case is required because
+            //
+            // - Using `ttnn::pad` changes the alignment on the tensor to its physical size, and the tensor spec
+            //   validation requires that `physical_shard_width % width_alignment == 0`
+            // - `physical_size.width() % page_shape.width() == 0`
+            if (input_tensor.logical_shape()[-1] % input_tensor_sharded_memory_config.shard_spec()->shape[1] != 0 &&
+                input_tensor.layout() == Layout::ROW_MAJOR &&
+                input_tensor_sharded_memory_config.memory_layout() != TensorMemoryLayout::HEIGHT_SHARDED) {
+                const auto& shard_spec = input_tensor_sharded_memory_config.shard_spec().value();
```
Comment (mstaletovicTT) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:912]: In this case, it seems like 0 padding is not applied at all ?

Comment (ctr-martemovTT) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:912]: Later on the code creates `tt::tt_metal::Alignment alignment{shard_spec.shape[0], shard_spec.shape[1]};` alignment based on the shard size, and uses it to create on-device tensor with the sharding. And it causes causes `TensorLayout::compute_physical_shape` to round up the logical shape to the shard dimensions automatically in https://github.com/tenstorrent/tt-metal/blob/b683ce5ce2afb2a383b13b71045ea9aa309a04ed/tt_metal/impl/tensor/spec/layout/tensor_layout.cpp#L384

So the padding caused by alignment to the shard size is the same that is computed for the `input_padded_shape`. 

This branch specifically follows the logic in the pre-existing code with similar logic in 

https://github.com/tenstorrent/tt-metal/blob/a9869cf90bbeee041c078568dac99f5bc3812c3b/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp#L873-L891

Comment (mstaletovicTT) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:912]: I understand that we "pad" in terms of declaring the correct physical shape, but before, this path did use to explicitly ttnn::pad, which does fill the host tensor with zeros. As a consequence, those zeros also fill the device tensor.

After this PR, it does not seem like that would hold, ie. the data in the final input tensor could be some random uninitialized variable that is not zeros

Comment (ctr-martemovTT) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:912]: @mstaletovicTT 
You are correct that the alignment-based tensor padding is going to use uninitialized values rather than explicit zeroes, but it is not a problem because this PR does not introduce *new* potential output: `input_tensor_on_device` branch already uses the same mechanism for padding the tensor, so the current implementation *already* has the path that contains unitialized data. 

Additionally, the read tensors explicitly handle zeroing out the data, e.g., so the case you are describing is already handled in the existing code. 

https://github.com/tenstorrent/tt-metal/blob/a9869cf90bbeee041c078568dac99f5bc3812c3b/ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/reader_conv_activations_padded_with_halo_3x3_weights_v2.cpp#L23

https://github.com/tenstorrent/tt-metal/blob/a9869cf90bbeee041c078568dac99f5bc3812c3b/ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/reader_conv_activations_padded_with_halo_3x3_weights_v2.cpp#L55-L57

**Review by mstaletovicTT — APPROVED** [review]
